File: lib/lab_sound/audio_scheduled_source_node.dart
import 'package:lab_sound_flutter/lab_sound/audio_node.dart';

import 'audio_context.dart';
import 'lab.dart';

class AudioScheduledSourceNode extends AudioNode {
  AudioScheduledSourceNode(AudioContext ctx, int nodeId): super(ctx, nodeId);
  bool get isPlayingOrScheduled => LabSound().AudioScheduledSourceNode_isPlayingOrScheduled(nodeId) > 0;
  stop({double when = 0}) => LabSound().AudioScheduledSourceNode_stop(nodeId, when);
  start({double when = 0}) => LabSound().AudioScheduledSourceNode_start(nodeId, when);
  bool get hasFinished => LabSound().AudioScheduledSourceNode_hasFinished(nodeId) > 0;
  int get startWhen => LabSound().AudioScheduledSourceNode_startWhen(nodeId);
  SchedulingState get playbackState => SchedulingState.values[LabSound().AudioScheduledSourceNode_playbackState(nodeId)];
}

File: bridging/AudioScheduledSourceNode.cpp
#include "dart_api/dart_api.h"
#include "LabSound/LabSound.h"
#include "KeepNode.cpp"
using namespace lab;

DART_EXPORT int AudioScheduledSourceNode_isPlayingOrScheduled(int nodeId) {
    return std::static_pointer_cast<AudioScheduledSourceNode>(audioNodes.find(nodeId)->second)->isPlayingOrScheduled();
}

DART_EXPORT void AudioScheduledSourceNode_stop(int nodeId, float when) {
    std::static_pointer_cast<AudioScheduledSourceNode>(audioNodes.find(nodeId)->second)->stop(when);
}

DART_EXPORT int AudioScheduledSourceNode_hasFinished(int nodeId) {
    return std::static_pointer_cast<AudioScheduledSourceNode>(audioNodes.find(nodeId)->second)->hasFinished();
}

DART_EXPORT uint64_t AudioScheduledSourceNode_startWhen(int nodeId) {
    return std::static_pointer_cast<AudioScheduledSourceNode>(audioNodes.find(nodeId)->second)->startWhen();
}

DART_EXPORT void AudioScheduledSourceNode_start(int nodeId, float when) {
    std::static_pointer_cast<AudioScheduledSourceNode>(audioNodes.find(nodeId)->second)->start(when);
}

DART_EXPORT int AudioScheduledSourceNode_playbackState(int nodeId) {
    return static_cast<int>(std::static_pointer_cast<AudioScheduledSourceNode>(audioNodes.find(nodeId)->second)->playbackState());
}
